diff --git a/drivers/pci/pci-sysfs.c b/drivers/pci/pci-sysfs.c
index e401f040f157..2369c139aa77 100644
--- a/drivers/pci/pci-sysfs.c
+++ b/drivers/pci/pci-sysfs.c
@@ -467,6 +467,33 @@ static ssize_t dev_rescan_store(struct device *dev,
 static struct device_attribute dev_attr_dev_rescan = __ATTR(rescan, 0200, NULL,
 							    dev_rescan_store);
 
+static ssize_t dev_bridge_rescan_store(struct device *dev,
+				struct device_attribute *attr, const char *buf,
+				size_t count)
+{
+	unsigned long val;
+	struct pci_dev *pdev = to_pci_dev(dev);
+	u32 buses;
+	pci_read_config_dword(pdev, PCI_PRIMARY_BUS, &buses);
+
+	if (kstrtoul(buf, 0, &val) < 0)
+		return -EINVAL;
+
+	if (val) {
+		pci_lock_rescan_remove();
+		pci_scan_bridge(pdev->bus,pdev,(buses >> 16) & 0xFF,0);
+		pci_assign_unassigned_bridge_resources(pdev);
+		pci_bus_add_devices(pdev->bus);
+		pci_unlock_rescan_remove();
+	}
+	return count;
+}
+
+static struct device_attribute dev_bridge_rescan_attr = __ATTR(bridge_rescan,
+							(S_IWUSR|S_IWGRP),
+							NULL, dev_bridge_rescan_store);
+
+
 static ssize_t remove_store(struct device *dev, struct device_attribute *attr,
 			    const char *buf, size_t count)
 {
@@ -1493,6 +1520,7 @@ static umode_t pci_dev_attrs_are_visible(struct kobject *kobj,
 static struct attribute *pci_dev_hp_attrs[] = {
 	&dev_attr_remove.attr,
 	&dev_attr_dev_rescan.attr,
+	&dev_bridge_rescan_attr.attr,
 	NULL,
 };
 
-- 
2.32.0

